Leaves the keys of hardware keyboards on iOS, iPadOS, Mac Catalyst and tvOS to GameController, through which the engine reads every key with its release, so SDL turns only the presses of remotes and game controllers into keys, whether a keyboard is connected or not, and hands UIKit the presses whose key the app drops through its event filter, such as the Menu press of the Apple TV remote while the app leaves the back button to the system. It also stops turning arrows and escape into key commands, and the swipes of the Apple TV remote into arrows, which the engine reads as the stick of the remote, and reads the Siri Remote that the joystick subsystem opens only where SDL builds that subsystem.

--- a/src/video/uikit/SDL_uikitview.m
+++ b/src/video/uikit/SDL_uikitview.m
@@ -38,8 +38,10 @@
 // The maximum number of mouse buttons we support
 #define MAX_MOUSE_BUTTONS 5
 
+#ifndef SDL_JOYSTICK_DISABLED
 // This is defined in SDL_sysjoystick.m
 extern int SDL_AppleTVRemoteOpenedAsJoystick;
+#endif
 
 @implementation SDL_uikitview
 {
@@ -57,25 +59,6 @@
 - (instancetype)initWithFrame:(CGRect)frame
 {
     if ((self = [super initWithFrame:frame])) {
-#ifdef SDL_PLATFORM_TVOS
-        // Apple TV Remote touchpad swipe gestures.
-        UISwipeGestureRecognizer *swipeUp = [[UISwipeGestureRecognizer alloc] initWithTarget:self action:@selector(swipeGesture:)];
-        swipeUp.direction = UISwipeGestureRecognizerDirectionUp;
-        [self addGestureRecognizer:swipeUp];
-
-        UISwipeGestureRecognizer *swipeDown = [[UISwipeGestureRecognizer alloc] initWithTarget:self action:@selector(swipeGesture:)];
-        swipeDown.direction = UISwipeGestureRecognizerDirectionDown;
-        [self addGestureRecognizer:swipeDown];
-
-        UISwipeGestureRecognizer *swipeLeft = [[UISwipeGestureRecognizer alloc] initWithTarget:self action:@selector(swipeGesture:)];
-        swipeLeft.direction = UISwipeGestureRecognizerDirectionLeft;
-        [self addGestureRecognizer:swipeLeft];
-
-        UISwipeGestureRecognizer *swipeRight = [[UISwipeGestureRecognizer alloc] initWithTarget:self action:@selector(swipeGesture:)];
-        swipeRight.direction = UISwipeGestureRecognizerDirectionRight;
-        [self addGestureRecognizer:swipeRight];
-#endif
-
 #if !defined(SDL_PLATFORM_TVOS)
         /* Pinch gestures */
         UIPinchGestureRecognizer *pinchGesture = [[UIPinchGestureRecognizer alloc] initWithTarget:self action:@selector(sdlPinchGesture:)];
@@ -520,7 +503,7 @@
 
     // Presses from Apple TV remote or game controller
     bool controller_opened = false;
-#ifdef SDL_PLATFORM_TVOS
+#if defined(SDL_PLATFORM_TVOS) && !defined(SDL_JOYSTICK_DISABLED)
     // tvOS doesn't send these for game controllers, but does for the Siri remote
     controller_opened = (SDL_AppleTVRemoteOpenedAsJoystick > 0);
 #else
@@ -556,42 +539,59 @@
     return SDL_SCANCODE_UNKNOWN;
 }
 
-- (void)pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+// The keys of a hardware keyboard reach the app through GameController, except the Escape key of tvOS, which makes the Menu press of the remote.
+- (BOOL)isKeyboardPress:(UIPress *)press
 {
-    if (!SDL_HasKeyboard()) {
-        for (UIPress *press in presses) {
-            SDL_Scancode scancode = [self scancodeFromPress:press];
-            SDL_SendKeyboardKey(UIKit_GetEventTimestamp([event timestamp]), SDL_GLOBAL_KEYBOARD_ID, 0, scancode, true);
+#ifdef SDL_PLATFORM_TVOS
+    return press.key != nil && press.type != UIPressTypeMenu;
+#else
+    return press.key != nil;
+#endif
+}
+
+// Sends the presses that are no keys of a keyboard as keys, and returns the ones whose key the app dropped through its event filter, which go on to UIKit, such as the Menu press of the remote that leaves the app.
+- (NSSet<UIPress *> *)sendPresses:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event down:(bool)down
+{
+    NSMutableSet<UIPress *> *dropped = [NSMutableSet set];
+    for (UIPress *press in presses) {
+        if ([self isKeyboardPress:press]) {
+            continue;
         }
+        SDL_Scancode scancode = [self scancodeFromPress:press];
+        if (!SDL_SendKeyboardKey(UIKit_GetEventTimestamp([event timestamp]), SDL_GLOBAL_KEYBOARD_ID, 0, scancode, down)) {
+            [dropped addObject:press];
+        }
     }
+    return dropped;
+}
+
+- (void)pressesBegan:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
+{
+    NSSet<UIPress *> *dropped = [self sendPresses:presses withEvent:event down:true];
     if (SDL_TextInputActive(sdlwindow)) {
         [super pressesBegan:presses withEvent:event];
+    } else if (dropped.count > 0) {
+        [super pressesBegan:dropped withEvent:event];
     }
 }
 
 - (void)pressesEnded:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
 {
-    if (!SDL_HasKeyboard()) {
-        for (UIPress *press in presses) {
-            SDL_Scancode scancode = [self scancodeFromPress:press];
-            SDL_SendKeyboardKey(UIKit_GetEventTimestamp([event timestamp]), SDL_GLOBAL_KEYBOARD_ID, 0, scancode, false);
-        }
-    }
+    NSSet<UIPress *> *dropped = [self sendPresses:presses withEvent:event down:false];
     if (SDL_TextInputActive(sdlwindow)) {
         [super pressesEnded:presses withEvent:event];
+    } else if (dropped.count > 0) {
+        [super pressesEnded:dropped withEvent:event];
     }
 }
 
 - (void)pressesCancelled:(NSSet<UIPress *> *)presses withEvent:(UIPressesEvent *)event
 {
-    if (!SDL_HasKeyboard()) {
-        for (UIPress *press in presses) {
-            SDL_Scancode scancode = [self scancodeFromPress:press];
-            SDL_SendKeyboardKey(UIKit_GetEventTimestamp([event timestamp]), SDL_GLOBAL_KEYBOARD_ID, 0, scancode, false);
-        }
-    }
+    NSSet<UIPress *> *dropped = [self sendPresses:presses withEvent:event down:false];
     if (SDL_TextInputActive(sdlwindow)) {
         [super pressesCancelled:presses withEvent:event];
+    } else if (dropped.count > 0) {
+        [super pressesCancelled:dropped withEvent:event];
     }
 }
 
@@ -603,34 +603,6 @@
     }
 }
 
-#ifdef SDL_PLATFORM_TVOS
-- (void)swipeGesture:(UISwipeGestureRecognizer *)gesture
-{
-    // Swipe gestures don't trigger begin states.
-    if (gesture.state == UIGestureRecognizerStateEnded) {
-        // tvOS doesn't send these for game controllers, but does for the Siri remote
-        bool controller_opened = (SDL_AppleTVRemoteOpenedAsJoystick > 0);
-        if (!controller_opened) {
-            /* Send arrow key presses for now, as we don't have an external API
-             * which better maps to swipe gestures. */
-            switch (gesture.direction) {
-            case UISwipeGestureRecognizerDirectionUp:
-                SDL_SendKeyboardKeyAutoRelease(0, SDL_SCANCODE_UP);
-                break;
-            case UISwipeGestureRecognizerDirectionDown:
-                SDL_SendKeyboardKeyAutoRelease(0, SDL_SCANCODE_DOWN);
-                break;
-            case UISwipeGestureRecognizerDirectionLeft:
-                SDL_SendKeyboardKeyAutoRelease(0, SDL_SCANCODE_LEFT);
-                break;
-            case UISwipeGestureRecognizerDirectionRight:
-                SDL_SendKeyboardKeyAutoRelease(0, SDL_SCANCODE_RIGHT);
-                break;
-            }
-        }
-    }
-}
-#endif // SDL_PLATFORM_TVOS
 
 @end
 
--- a/src/video/uikit/SDL_uikitviewcontroller.m
+++ b/src/video/uikit/SDL_uikitviewcontroller.m
@@ -306,39 +306,6 @@
                  object:nil];
 }
 
-- (NSArray *)keyCommands
-{
-    NSMutableArray *commands = [[NSMutableArray alloc] init];
-    [commands addObject:[UIKeyCommand keyCommandWithInput:UIKeyInputUpArrow modifierFlags:kNilOptions action:@selector(handleCommand:)]];
-    [commands addObject:[UIKeyCommand keyCommandWithInput:UIKeyInputDownArrow modifierFlags:kNilOptions action:@selector(handleCommand:)]];
-    [commands addObject:[UIKeyCommand keyCommandWithInput:UIKeyInputLeftArrow modifierFlags:kNilOptions action:@selector(handleCommand:)]];
-    [commands addObject:[UIKeyCommand keyCommandWithInput:UIKeyInputRightArrow modifierFlags:kNilOptions action:@selector(handleCommand:)]];
-    [commands addObject:[UIKeyCommand keyCommandWithInput:UIKeyInputEscape modifierFlags:kNilOptions action:@selector(handleCommand:)]];
-    return [NSArray arrayWithArray:commands];
-}
-
-- (void)handleCommand:(UIKeyCommand *)keyCommand
-{
-    SDL_Scancode scancode = SDL_SCANCODE_UNKNOWN;
-    NSString *input = keyCommand.input;
-
-    if (input == UIKeyInputUpArrow) {
-        scancode = SDL_SCANCODE_UP;
-    } else if (input == UIKeyInputDownArrow) {
-        scancode = SDL_SCANCODE_DOWN;
-    } else if (input == UIKeyInputLeftArrow) {
-        scancode = SDL_SCANCODE_LEFT;
-    } else if (input == UIKeyInputRightArrow) {
-        scancode = SDL_SCANCODE_RIGHT;
-    } else if (input == UIKeyInputEscape) {
-        scancode = SDL_SCANCODE_ESCAPE;
-    }
-
-    if (scancode != SDL_SCANCODE_UNKNOWN) {
-        SDL_SendKeyboardKeyAutoRelease(0, scancode);
-    }
-}
-
 - (void)setView:(UIView *)view
 {
     [super setView:view];
